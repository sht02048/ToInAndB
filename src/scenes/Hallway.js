import Renderer from "../graphics/Renderer";
import Mini from "../entities/enemies/Mini";
import Cannon from "../entities/enemies/Cannon";
import Danger from "../entities/enemies/Danger";

class Hallway extends Renderer {
  #isDone = true;
  #miniWidth = 44;

  constructor() {
    super();

    this.mini = new Mini(this.canvasWidth / 2 - this.#miniWidth / 2, -20);
    this.cannonList = [
      new Cannon({ x: this.minX + 100, y: -100 }),
      new Cannon({ x: this.maxX - 100, y: -100 }),
    ];
    this.dangerList = Array.from({ length: 3 }, (_, index) => [
      new Danger({
        x: this.canvasWidth / 2 - 300,
        y: index * -100 - 100,
        isLeft: true,
      }),
      new Danger({
        x: this.canvasWidth / 2 + 300,
        y: index * -100 - 100,
        isLeft: false,
      }),
    ]).flat();
  }

  update() {
    this.mini.update();
    this.cannonList.forEach((cannon) => cannon.update());
    this.dangerList.forEach((danger) => danger.update());
  }

  render() {
    this.mini.render();
    this.cannonList.forEach((cannon) => cannon.render());
    this.dangerList.forEach((danger) => danger.render());
  }

  setTarget(player) {
    this.mini.setTargetList(player);
    this.cannonList.forEach((cannon) => cannon.setTargetList(player));
    this.dangerList.forEach((danger) => danger.setTargetList(player));
  }

  setSceneTargetList() {
    const sceneTargetList = [this.mini];

    this.cannonList.forEach((cannon) => sceneTargetList.push(cannon));
    this.dangerList.forEach((danger) => sceneTargetList.push(danger));

    return sceneTargetList;
  }

  checkSceneStatus() {
    this.#isDone = true;

    if (!this.mini.isDestroyed && !this.mini.isVanished) {
      this.#isDone = false;
    }

    this.cannonList.forEach((cannon) => {
      if (!cannon.isDestroyed && !cannon.isVanished) {
        this.#isDone = false;
      }
    });

    this.dangerList.forEach((danger) => {
      if (!danger.isDestroyed && !danger.isVanished) {
        this.#isDone = false;
      }
    });

    return this.#isDone;
  }
}

export default Hallway;
