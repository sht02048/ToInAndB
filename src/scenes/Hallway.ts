import Renderer from "../graphics/Renderer";

import Player from "../entities/Player";
import Mini from "../entities/enemies/Mini";
import Cannon from "../entities/enemies/Cannon";
import Danger from "../entities/enemies/Danger";

class Hallway extends Renderer {
  #isDone = true;
  #miniWidth = 44;

  private mini: Mini;
  private cannonList: Cannon[];
  private dangerList: Danger[];
  private shouldBeDisplayed: boolean;
  private hasStarted: boolean;

  constructor() {
    super();

    this.mini = new Mini(this.canvasWidth / 2 - this.#miniWidth / 2, -45);
    this.cannonList = [
      new Cannon({ x: this.minX + 100, y: -100 }),
      new Cannon({ x: this.maxX - 100, y: -100 }),
    ];
    this.dangerList = Array.from({ length: 3 }, (_, index) => [
      new Danger({
        x: this.canvasWidth / 2 - 300,
        y: index * -100 - 100,
        isLeft: true,
      }),
      new Danger({
        x: this.canvasWidth / 2 + 300,
        y: index * -100 - 100,
        isLeft: false,
      }),
    ]).flat();

    this.shouldBeDisplayed = false;
    this.hasStarted = false;
  }

  update() {
    this.mini.update();
    this.cannonList.forEach((cannon) => cannon.update());
    this.dangerList.forEach((danger) => danger.update());
  }

  render() {
    this.mini.render();
    this.cannonList.forEach((cannon) => cannon.render());
    this.dangerList.forEach((danger) => danger.render());
  }

  setTarget(player: Player) {
    this.mini.setTargetList(player);
    this.cannonList.forEach((cannon) => cannon.setTargetList(player));
    this.dangerList.forEach((danger) => danger.setTargetList(player));
  }

  setSceneTargetList() {
    const sceneTargetList = [this.mini, ...this.cannonList, ...this.dangerList];

    return sceneTargetList;
  }

  checkSceneStatus() {
    this.#isDone = true;

    if (!this.mini.isDestroyed && !this.mini.isVanished) {
      this.#isDone = false;
    }

    this.cannonList.forEach((cannon) => {
      if (!cannon.isDestroyed && !cannon.isVanished) {
        this.#isDone = false;
      }
    });

    this.dangerList.forEach((danger) => {
      if (!danger.isDestroyed && !danger.isVanished) {
        this.#isDone = false;
      }
    });

    return this.#isDone;
  }
}

export default Hallway;
