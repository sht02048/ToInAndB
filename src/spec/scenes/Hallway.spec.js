import Hallway from "../../scenes/Hallway";

describe("Hallway", () => {
  const mainCanvas = document.createElement("canvas");
  const introCanvas = document.createElement("canvas");
  mainCanvas.id = "main-canvas";
  introCanvas.id = "intro-canvas";

  document.body.appendChild(mainCanvas);
  document.body.appendChild(introCanvas);

  let hallway;

  beforeEach(() => {
    hallway = new Hallway();
  });

  it("should have proper functions", () => {
    expect(hallway.update).to.be.a("function");
    expect(hallway.render).to.be.a("function");
    expect(hallway.setTarget).to.be.a("function");
    expect(hallway.setSceneTargetList).to.be.a("function");
    expect(hallway.checkSceneStatus).to.be.a("function");
  });

  it("should render mini, cannon and danger", () => {
    hallway.render();

    const ctx = mainCanvas.getContext("2d");
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });

  it("should update mini, cannon and danger", () => {
    hallway.update();
    hallway.render();

    const ctx = mainCanvas.getContext("2d");
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });

  it("should check if scene is done", () => {
    hallway.mini.isDestroyed = true;
    hallway.cannonList.forEach((cannon) => (cannon.isDestroyed = true));
    hallway.dangerList.forEach((danger) => (danger.isDestroyed = true));

    expect(hallway.checkSceneStatus()).to.equal(true);
  });
});
