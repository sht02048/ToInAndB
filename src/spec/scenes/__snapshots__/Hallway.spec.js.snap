// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Hallway > should render mini, cannon and danger 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 128,
      "dy": -45,
      "img": <img
        src="/src/assets/sprites/enemies/mini.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 200,
      "dy": -100,
      "img": <img
        src="/src/assets/sprites/enemies/cannon.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 55,
      "dy": -100,
      "img": <img
        src="/src/assets/sprites/enemies/cannon.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -150,
      "dy": -100,
      "img": <img
        src="/src/assets/sprites/enemies/danger.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 450,
      "dy": -100,
      "img": <img
        src="/src/assets/sprites/enemies/danger.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -150,
      "dy": -200,
      "img": <img
        src="/src/assets/sprites/enemies/danger.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 450,
      "dy": -200,
      "img": <img
        src="/src/assets/sprites/enemies/danger.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -150,
      "dy": -300,
      "img": <img
        src="/src/assets/sprites/enemies/danger.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 450,
      "dy": -300,
      "img": <img
        src="/src/assets/sprites/enemies/danger.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`Hallway > should update mini, cannon and danger 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 128,
      "dy": -45,
      "img": <img
        src="/src/assets/sprites/enemies/mini.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 200,
      "dy": -100,
      "img": <img
        src="/src/assets/sprites/enemies/cannon.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 55,
      "dy": -100,
      "img": <img
        src="/src/assets/sprites/enemies/cannon.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -150,
      "dy": -100,
      "img": <img
        src="/src/assets/sprites/enemies/danger.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 450,
      "dy": -100,
      "img": <img
        src="/src/assets/sprites/enemies/danger.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -150,
      "dy": -200,
      "img": <img
        src="/src/assets/sprites/enemies/danger.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 450,
      "dy": -200,
      "img": <img
        src="/src/assets/sprites/enemies/danger.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -150,
      "dy": -300,
      "img": <img
        src="/src/assets/sprites/enemies/danger.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 450,
      "dy": -300,
      "img": <img
        src="/src/assets/sprites/enemies/danger.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 128,
      "dy": -43,
      "img": <img
        src="/src/assets/sprites/enemies/mini.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 200,
      "dy": -98,
      "img": <img
        src="/src/assets/sprites/enemies/cannon.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 55,
      "dy": -98,
      "img": <img
        src="/src/assets/sprites/enemies/cannon.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
